### PR TITLE
Segfault after obspy.read

### PR DESCRIPTION
**Pull request by krischer**
Hi devs, running the QC we are sometimes running in to a segfault reading the record headers of some files. It's been quite a challenge but I traced it back to the `obspy.read` subroutine... it really blows my mind.

    from obspy import read
    from obspy.io.mseed.util import get_flags

    # Reading the file before getting the records causes a segfault
    # when getting header flags.
    read("broken_records.mseed")
    get_flags("broken_records.mseed")

    >>> Warning: Number of blockettes in fixed header (2) does not match the number parsed (1)
    >>> Segmentation fault (core dumped)

When we do not read the file first the segmentation fault does not occur.:

    from obspy import read
    from obspy.io.mseed.util import get_flags

    get_flags("broken_records.mseed")

    >>> Warning: Number of blockettes in fixed header (2) does not match the number parsed (1)
    >>> Warning: Number of blockettes in fixed header (2) does not match the number parsed (1)
    >>> ... identical for all 16 records

Here is the file: `ftp://www.orfeus-eu.org/pub/outgoing/broken_records.mseed` (8 kb).

Thanks for the help and hope you understand this behaviour better than me..

### CODE REVIEW

**Comment by krischer** [comment]
This is now A PR fixing the issue.

The problem was that the call to `read_mseed()` hooked up libmseed's logging to some Python callbacks. Leaving the read function the Python functions were garbage collected and a call the next time naturally resulted in a segfault.

There is now a wrapper around the ctypes lib to always setup the logging so this can no longer happen and it is completely transparent to anyone directly using libmseed from Python.

Does not fix #1728 which is another issue.

**Comment by krischer** [comment]
IMHO ready to go.

File: obspy/io/mseed/headers.py [hunk 1]
```diff
@@ -581,11 +497,11 @@ class MSTraceSeg(C.Structure):
     ('samprate', C.c_double),         # Nominal sample rate (Hz)
     ('samplecnt', C.c_int64),         # Number of samples in trace coverage
     ('datasamples', C.c_void_p),      # Data samples, 'numsamples' of type
-                                      # 'sampletype'
+    # 'sampletype'
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: This is wrapped from the previous line; looks pretty weird unindented like this.

File: obspy/io/mseed/headers.py [hunk 2]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
+    def __init__(self, lib):
+        self.lib = lib
+        self.verbose = True
+
+    def __getattr__(self, item):
+        func = getattr(self.lib, item)
+
+        def _wrapper(*args):
+            # Collect exceptions. They cannot be raised in the callback as
+            # they could never be caught then. They are collected an raised
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: s/an/and/

File: obspy/io/mseed/headers.py [hunk 3]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
+                msg = msg.decode()
+                if msg.startswith("ERROR: "):
+                    msg = msg[7:].strip()
+                    _errs.append(msg)
+                if msg.startswith("INFO: "):
+                    msg = msg[6:].strip()
+                    _warns.append(msg)
+
+            diag_print = \
+                C.CFUNCTYPE(C.c_void_p, C.c_char_p)(log_error_or_warning)
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: Return type is `None`, not `c_void_p`.

File: obspy/io/mseed/headers.py [hunk 4]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
+
+            def log_message(msg):
+                if self.verbose:
+                    print(msg[6:].strip())
+            log_print = C.CFUNCTYPE(C.c_void_p, C.c_char_p)(log_message)
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: Return type is `None`, not `c_void_p`.

File: obspy/io/mseed/headers.py [hunk 5]
```diff
@@ -699,12 +615,98 @@ class LinkedIDList(C.Structure):
...
+##########################################################################
+
+__clibmseed = _load_cdll("mseed")
+
+# Declare function of libmseed library, argument parsing
+__clibmseed.mst_init.argtypes = [C.POINTER(MSTrace)]
+__clibmseed.mst_init.restype = C.POINTER(MSTrace)
+
+__clibmseed.mst_free.argtypes = [C.POINTER(C.POINTER(MSTrace))]
+__clibmseed.mst_free.restype = C.c_void_p
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`

File: obspy/io/mseed/headers.py [hunk 6]
```diff
@@ -699,12 +615,98 @@ class LinkedIDList(C.Structure):
...
+
+__clibmseed.mst_initgroup.argtypes = [C.POINTER(MSTraceGroup)]
+__clibmseed.mst_initgroup.restype = C.POINTER(MSTraceGroup)
+
+__clibmseed.mst_freegroup.argtypes = [C.POINTER(C.POINTER(MSTraceGroup))]
+__clibmseed.mst_freegroup.restype = C.c_void_p
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`

File: obspy/io/mseed/headers.py [hunk 7]
```diff
@@ -699,12 +615,98 @@ class LinkedIDList(C.Structure):
...
+    np.ctypeslib.ndpointer(dtype=np.int32, ndim=1,
+                           flags=native_str('C_CONTIGUOUS')),
+    C.c_int, C.c_char_p, C.c_int]
+__clibmseed.msr_decode_steim1.restype = C.c_int
+
+# tricky, C.POINTER(C.c_char) is a pointer to single character fields
+# this is completely different to C.c_char_p which is a string
+__clibmseed.mst_packgroup.argtypes = [
+    C.POINTER(MSTraceGroup), C.CFUNCTYPE(
+        C.c_void_p, C.POINTER(C.c_char), C.c_int, C.c_void_p),
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`, not `c_void_p` (the first time).

File: obspy/io/mseed/headers.py [hunk 8]
```diff
@@ -699,12 +615,98 @@ class LinkedIDList(C.Structure):
...
+    C.c_void_p, C.c_int, C.c_short, C.c_short, C.POINTER(C.c_int), C.c_short,
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `C.POINTER(C.c_int64)`, actually

File: obspy/io/mseed/headers.py [hunk 9]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
 
-clibmseed.readMSEEDBuffer.restype = C.POINTER(LinkedIDList)
 
-clibmseed.msr_free.argtypes = [C.POINTER(C.POINTER(MSRecord))]
-clibmseed.msr_free.restype = C.c_void_p
+__clibmseed.setupLogging.argtpyes = [
+    C.c_int8,
+    C.CFUNCTYPE(C.c_void_p, C.c_char_p),
+    C.CFUNCTYPE(C.c_void_p, C.c_char_p)]
+__clibmseed.setupLogging.restype = C.c_void_p
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`

File: obspy/io/mseed/headers.py [hunk 10]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
 
-clibmseed.mstl_init.restype = C.POINTER(MSTraceList)
-clibmseed.mstl_free.argtypes = [C.POINTER(C.POINTER(MSTraceList)), C.c_int]
 
+__clibmseed.msr_free.argtypes = [C.POINTER(C.POINTER(MSRecord))]
+__clibmseed.msr_free.restype = C.c_void_p
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`

File: obspy/io/mseed/headers.py [hunk 11]
```diff
@@ -714,46 +716,103 @@ class LinkedIDList(C.Structure):
...
-clibmseed.lil_free.restype = C.c_void_p
 
+__clibmseed.mstl_init.restype = C.POINTER(MSTraceList)
+__clibmseed.mstl_free.argtypes = [C.POINTER(C.POINTER(MSTraceList)), C.c_int]
 
-clibmseed.allocate_bytes.argtypes = (C.c_int,)
-clibmseed.allocate_bytes.restype = C.c_void_p
 
+__clibmseed.lil_free.argtypes = [C.POINTER(LinkedIDList)]
+__clibmseed.lil_free.restype = C.c_void_p
```
Comment (QuLogic) [inline obspy/io/mseed/headers.py]: `None`

**Comment by megies** [comment]
Thanks for the review @QuLogic, I've opened a follow up issue to make sure your comments don't go unnoticed..

**Pull request merged**